pkg/apis/pipeline/v1beta1/pipelinerun_conversion.go

```diff
@@ -20,23 +20,205 @@ import (
+		newTimeouts := &TimeoutFields{}
+		newTimeouts.convertFrom(ctx, *source.Timeouts)
+		prs.Timeouts = newTimeouts
+	}
+	prs.PodTemplate = source.PodTemplate
+	prs.Workspaces = nil
+	for _, w := range source.Workspaces {
+		new := WorkspaceBinding{}
+		new.convertFrom(ctx, w)
+		prs.Workspaces = append(prs.Workspaces, new)
+	}
+	prs.TaskRunSpecs = nil
+	for _, trs := range source.TaskRunSpecs {
+		new := PipelineTaskRunSpec{}
+		new.convertFrom(ctx, trs)
+		prs.TaskRunSpecs = append(prs.TaskRunSpecs, new)
+	}
+	return nil
+}
+
+func (tf TimeoutFields) convertTo(ctx context.Context, sink *v1.TimeoutFields) {
+	sink.Pipeline = tf.Pipeline
+	sink.Tasks = tf.Tasks
+	sink.Finally = tf.Finally
+}
+
+func (tf *TimeoutFields) convertFrom(ctx context.Context, source v1.TimeoutFields) {
+	tf.Pipeline = source.Pipeline
+	tf.Tasks = source.Tasks
+	tf.Finally = source.Finally
+}
+
+func (ptrs PipelineTaskRunSpec) convertTo(ctx context.Context, sink *v1.PipelineTaskRunSpec) {
+	sink.PipelineTaskName = ptrs.PipelineTaskName
+	sink.ServiceAccountName = ptrs.TaskServiceAccountName
+	sink.PodTemplate = ptrs.TaskPodTemplate
+	sink.StepOverrides = nil
+	for _, so := range ptrs.StepOverrides {
+		new := v1.TaskRunStepOverride{}
+		so.convertTo(ctx, &new)
+		sink.StepOverrides = append(sink.StepOverrides, new)
+	}
+	sink.SidecarOverrides = nil
+	for _, so := range ptrs.SidecarOverrides {
+		new := v1.TaskRunSidecarOverride{}
+		so.convertTo(ctx, &new)
+		sink.SidecarOverrides = append(sink.SidecarOverrides, new)
+	}
+	if ptrs.Metadata != nil {
+		sink.Metadata = &v1.PipelineTaskMetadata{}
+		ptrs.Metadata.convertTo(ctx, sink.Metadata)
+	}
+	sink.ComputeResources = ptrs.ComputeResources
+}
+
+func (ptrs *PipelineTaskRunSpec) convertFrom(ctx context.Context, source v1.PipelineTaskRunSpec) {
+	ptrs.PipelineTaskName = source.PipelineTaskName
+	ptrs.TaskServiceAccountName = source.ServiceAccountName
+	ptrs.TaskPodTemplate = source.PodTemplate
+	ptrs.StepOverrides = nil
+	for _, so := range source.StepOverrides {
+		new := TaskRunStepOverride{}
+		new.convertFrom(ctx, so)
+		ptrs.StepOverrides = append(ptrs.StepOverrides, new)
+	}
+	ptrs.SidecarOverrides = nil
+	for _, so := range source.SidecarOverrides {
+		new := TaskRunSidecarOverride{}
+		new.convertFrom(ctx, so)
+		ptrs.SidecarOverrides = append(ptrs.SidecarOverrides, new)
+	}
+	if source.Metadata != nil {
+		newMetadata := PipelineTaskMetadata{}
+		newMetadata.convertFrom(ctx, *source.Metadata)
+		ptrs.Metadata = &newMetadata
+	}
+	ptrs.ComputeResources = source.ComputeResources
+}
+
+func serializePipelineRunResources(meta *metav1.ObjectMeta, spec *PipelineRunSpec) error {
+	if spec.Resources == nil {
 		return nil
 	}
-	return fmt.Errorf("v1beta1 is the highest known version, got: %T", source)
+	return version.SerializeToMetadata(meta, spec.Resources, resourcesAnnotationKey)
+}
+
+func deserializePipelineRunResources(meta *metav1.ObjectMeta, spec *PipelineRunSpec) error {
+	resources := &[]PipelineResourceBinding{}
```

Comment:
I'd prefer `resources := []PipelineResourceBinding{}`-- I don't really have a good reason, but the dereference on line 220 made me check whether it was possible to be a nil pointer, and if resources isn't a pointer that mental check doesn't have to happen.